src/sampling.jl

```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
 end
 seqsample_c!(a::AbstractArray, x::AbstractArray) = seqsample_c!(Random.GLOBAL_RNG, a, x)
 
-## TODO: implement Algorithm D (page 716 - 717)
+"""
+    seqsample_d!([rng], a::AbstractArray, x::AbstractArray)
+Random subsequence sampling using algorithm D described in the following paper (page 716-17):
+Jeffrey Scott Vitter. "Faster Methods for Random Sampling". Communications of the ACM,
+27 (7), July 1984.
+This algorithm consumes ``O(k)`` random numbers, with `k=length(x)`.
+The outputs are ordered.
+"""
+function seqsample_d!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
+    N = length(a)
+    n = length(x)
+    n <= N || error("length(x) should not exceed length(a)")
+
+    i = 0
+    j = 0
+
+    vprime = exp(log(rand(rng)) / n)
```

Comment:
```suggestion
    vprime = exp(-randexp(rng)/n)
```

Reply:
Are you sure about this change ?

Reply:
The inverse of the cdf of an exponential random variable `randexp()` is `x = –log(1–u)`. Therefore `-log(1-rand())` and `-log(rand())` have the same distribution as `randexp`, and `randexp` has the better resolution. Not sure what matters here.